### DISCUSSION TITLE
Can there be support for creating views in sqlite_orm?

### DISCUSSION THREAD

**Discussion by juandent** [fnc12/sqlite_orm#968] [Ideas]
Right now one can create a table in C++ with the make_storage function and afterwards make the table a view in any sqlite client like sqlitestudio. Sync_schema() fails to preserve data when columns are added and fails to change CHECK clauses in the database so it is not working as one would like (I understand there is complexity there...). I don't know how hard can it be to add a make_view() function to use inside make_storage, but I feel it would greatly add to sqlite_orm's functionality! In my opinion this should have high priority...

**Comment by fnc12** [comment]
adding views is one of the most impressive features. It exists in TODO list https://github.com/fnc12/sqlite_orm/blob/master/TODO.md. One issue is API. We need to make API which will allow performing any operation with view available in SQLite and (desired) keep compile time checks just like it is done with table. Do you have any idea?

**Comment by juandent** [comment]
Yes, my proposal is that we would use storage to create the SQL and then define the view by passing the statement, in this way we would be using all the serialization and we could place any statement no matter how complex in the view definition, something like:

```C++
auto statement = 
        storage.prepare(select(columns(&Employee::m_ename, &Employee::m_job), where(in(&Employee::m_deptno, { 10,20 }))));
	make_view("employees_in_deptnos_10_or_20", as(statement)); // this would go inside make_storage() 
```
Couldn't it be as simple as this?

**Reply by fnc12** [reply]
ok but how then we will make a reference to a column of such a view when we need to select this column?

**Reply by juandent** [reply]
we could create a struct with the view's parameters and make that part of the view definition like this:

```C++
	struct EmployeesInDept   // this would be the interface of the view
	{
		std::string m_ename;
		std::string m_job;
	};
	auto statement = storage.prepare(select(columns(&Employee::m_ename, &Employee::m_job), where(in(&Employee::m_deptno, { 10,20 }))));

	make_view("employees_in_dept_10_or_20", &EmployeesInDept::m_ename, &EmployeesInDept::m_job, as(statement));
```

**Reply by fnc12** [reply]
Ok. I think it would be better specifying view schema inside `make_storage` call just like it is done with other schema entities like tables, indexes and triggers. Like this:
```c++
auto storage = make_storage("...",
    make_table(...),
    make_index(...),
    make_trigger(...),
    make_view(...)
);
```

**Reply by juandent** [reply]
Sure that would work better

**Reply by juandent** [reply]
but it seems we have all the tough parts already at our disposal? does this mean implementing the feature would not be as big a deal?

**Reply by fnc12** [reply]
yes looks like you are right

**Reply by juandent** [reply]
does this make your plans to add views in the near future a reality?

**Reply by fnc12** [reply]
yes

**Comment by juandent** [comment]
I would love to create examples of views in the Guide!! 😁 BTW how did you liked it?

**Reply by juandent** [reply]
I think the make_view could look like this:

```C++
make_view("employees_in_dept_10_or_20", 
       make_parameter("ename", &EmployeesInDept::m_ename),
       make_parameter( "job", &EmployeesInDept::m_job), 
       as(statement));
```

**Reply by fnc12** [reply]
I'd skip `statement` cause `statement` is created by `sqlite3_prepare_v2` which is not required during storage creation:
```c++
make_view("employees_in_dept_10_or_20",
    make_column("ename", &EmployeesInDept::m_ename),
    make_column("job", &EmployeesInDept::m_job),
    as(select(columns(&Employee::m_ename, &Employee::m_job), where(in(&Employee::m_deptno, { 10,20 })))
```
It will make a static check that number of columns provided with `make_column` functions is equal to amount of columns inside `select`

**Comment by trueqbit** [comment]
Working on this feature in experimental branch [sql-view](../tree/experimental/sql-view).

**Comment by trueqbit** [comment]
PR #1481 merged...

There's a "high earning employees" example in the repository: [examples/view.cpp](/fnc12/sqlite_orm/blob/dev/examples/view.cpp).
